File: src/main/rs/emulate/shared/prop/PropertyType.java
package rs.emulate.shared.prop;

import rs.emulate.util.StringUtils;

/**
 * A type of a property. Should <strong>only</strong> be implemented by enumerators (excluding the existing
 * {@link DynamicPropertyType} class).
 * 
 * @author Major
 */
public interface PropertyType {

	/**
	 * Gets the name of this PropertyType, capitalised and with underscores ('_') replaced with spaces (' ').
	 * 
	 * @return The formatted name.
	 */
	default public String formattedName() {
		return StringUtils.capitalise(name().replace('_', ' '));
	}

	/**
	 * Gets the name of this PropertyType, as a String.
	 * 
	 * @return The name.
	 */
	public String name();

}

File: util/src/main/kotlin/rs/emulate/shared/property/PropertyType.kt
package rs.emulate.shared.property

/**
 * A type of a property.
 */
interface PropertyType {

    /**
     * Gets the name of this PropertyType, capitalised and with underscores ('_') replaced with spaces (' ').
     */
    fun formattedName(): String {
        return capitalise(name, split = "_")
    }

    /**
     * The name of the property.
     */
    val name: String

    companion object {

        /**
         * Returns a new string consisting of the characters in the specified String, with the first letter of each word
         * capitalised, and the rest lower-case.
         */
        private fun capitalise(string: String, split: String = " "): String {
            return string.split(split.toRegex())
                .dropLastWhile(String::isEmpty)
                .joinToString(" ") {
                    it[0].toUpperCase() + it.substring(1).toLowerCase()
                }
        }

    }

}
